### PR TITLE
Introduce a small delay in when sending multiple http_probe.requests_…

### PR DESCRIPTION
**Pull request by haraldschioberg**
…per_probe

This is in the same spirit as interval_between_targets_ms but
it's using a predicatable delay. Since those sub-probes are
(a) run once per probe
(b) in a predictable for-loop
there is no reason to have a random jitter, just space them all out to prevent them bunching up.

### CODE REVIEW

**Review by manugarg — COMMENTED** [review]
Thanks for the PR @haraldschioberg :)

I added a couple of inline comments, but other than that we'll need to do a couple of more things:
 = We'll need to verify that requests interval is compatible with the probe interval, to make sure that all requests can be done within one probe interval.
 = We'll need to move timeout handling to doHTTPRequest so that each request gets the same time to execute.

I'll do something like this in the Init() function of the probe:
```
if p.c.GetRequestsIntervalMsec() * p.c.GetRequestsPerProbe() > (p.opts.Interval - p.opts.Timeout).Milliseconds() {
  // Return error.
  // We have a problem. We cannot make all requests within one probe interval.
}
```

I'll also add a comment to the config proto file:
```
// ...
// This value should be less than (interval - timeout) / requests_per_probe. This
// is to ensure that all requests are executed within one probe interval and all of
// them get sufficient time. For example, if probe interval is 2s and timeout is 1s,
// and requests_per_probe is 10,  requests_interval_msec should be less than 10ms.
```

File: probes/http/proto/config.proto [hunk 1]
```diff
@@ -94,6 +94,11 @@ message ProbeConf {
   // requests per probe, "total" counter will be incremented by 2.
   optional int32 requests_per_probe = 98 [default = 1];
 
+  // Delay between request per probe
+  // Delay between the individual "requests_per_probe". This is ignored
+  // when requests_per_probe is "1".
+  optional int32 delay_requests_per_probe_ms = 100 [default = 1];
+
```
[the remark this answers is not in the document]
Comment (haraldschioberg) [reply probes/http/proto/config.proto]: Done.

File: probes/http/http.go [hunk 2]
```diff
@@ -346,11 +346,14 @@ func (p *Probe) runProbe(ctx context.Context, target endpoint.Endpoint, req *htt
 
 	wg := sync.WaitGroup{}
 	for numReq := int32(0); numReq < p.c.GetRequestsPerProbe(); numReq++ {
+		initialDelay := time.Duration(numReq*p.c.GetDelayRequestsPerProbeMs()) * time.Millisecond
 		wg.Add(1)
-		go func(req *http.Request, targetName string, result *probeResult) {
+		go func(initial_delay time.Duration, req *http.Request, targetName string, result *probeResult) {
```
[the remark this answers is not in the document]
Comment (haraldschioberg) [reply probes/http/http.go]: Done, thanks!

**Comment by haraldschioberg** [comment]
Moved the delay logic around a bit. 

What I am trying to do is to use an http connection for each repeat-probe: my target is loadbalanced, and I am trying to get a representative sample of all target instances. So I need a large number of keepalive http connections. I am trying to get to 99.9% latency metrics for the system, and there are ~thousands of instances: that means ~thousands of http connections.

**Review by manugarg — COMMENTED** [review]
Thanks for the updates, @haraldschioberg.

**Comment by haraldschioberg** [comment]
This is a rather naive and minimum effort attempt to hack better parallelism into cloudprober. The dynamic binding of probe requests opportunistically picking an open httptrace means that its unpredictable how many probe requests reuse the same connection.
A much better solution would be to, for each requests_per_probe to open a long-lived goroutine, which each open a dedicated http connection, then block on a timer.trigger channel, and to return results over a results channel for merging in the main probe.
The main routine just needs to send slightly spaced out triggers to each channel, and wait for results, also gets rid of the mutex on the results.

With these parameters:

hostname:www.google.com 
relative_url: /generate_204 
requests_per_probe: <100-1000>
interval_msec: <2000-10000> 
keepalive: true
requests_inteval_msec: 2

I've seen as little as 5 connection opened when sleeping outside doHttpRequest, and never more than ~ (0.7 * requests_per_probe) when sleeping after acquiring a connection from httptrace.

**Comment by manugarg** [comment]
@haraldschioberg I was wondering you wanted to give it a try again. I think the first change (with some comments addressed) should work with the HEAD now.

**Comment by manugarg** [comment]
Committed this pull request through: https://github.com/cloudprober/cloudprober/commit/77f01678fd8f674575c0984ca3ad5dd2cb78f200.

Closing this now. Please let me know if this doesn't look good.

**Pull request closed without merging**